### PR TITLE
[KEYCLOAK-5628] Added endpoints for applications

### PR DESCRIPTION
**Pull request by douglaspalmer**

### CODE REVIEW

File: services/src/main/java/org/keycloak/services/resources/account/AccountRestService.java [hunk 1]
```diff
@@ -464,6 +463,15 @@ private UserConsentModel createConsent(ClientModel client, ConsentRepresentation
         return consent;
     }
 
+    @Path("/applications")
```
Comment (abstractj) [inline services/src/main/java/org/keycloak/services/resources/account/AccountRestService.java]: @douglaspalmer while I was reviewing your PR, just reading the sources I found out something similar to what you implemented https://github.com/keycloak/keycloak/blob/c6c5a427537b868640d31310505f7a6411bd861c/services/src/main/java/org/keycloak/services/resources/account/AccountRestService.java#L251

Do we really need the code for applications in both places? Maybe we could merge the logic of this code in a single place?

File: services/src/main/java/org/keycloak/services/resources/account/AccountRestService.java [hunk 2]
```diff
@@ -464,6 +463,15 @@ private UserConsentModel createConsent(ClientModel client, ConsentRepresentation
...
+    public ApplicationResource applications() {
+        checkAccountApiEnabled();
+//        auth.requireOneOf(AccountRoles.MANAGE_ACCOUNT, AccountRoles.VIEW_PROFILE);
```
Comment (abstractj) [inline services/src/main/java/org/keycloak/services/resources/account/AccountRestService.java]: @douglaspalmer is this temporary?

Comment (abstractj) [reply services/src/main/java/org/keycloak/services/resources/account/AccountRestService.java]: Btw I commented out the following snippet:

```
@Path("/applications")
    public ApplicationResource applications() {
        checkAccountApiEnabled();
//        auth.requireOneOf(AccountRoles.MANAGE_ACCOUNT, AccountRoles.VIEW_PROFILE);
        return new ApplicationResource(session, user, auth, request);
    }
```

And I could confirm that the endpoint responding is https://github.com/keycloak/keycloak/blob/c6c5a427537b868640d31310505f7a6411bd861c/services/src/main/java/org/keycloak/services/resources/account/AccountRestService.java#L251

File: services/src/main/java/org/keycloak/services/resources/account/ApplicationResource.java [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+
+        private boolean internal;
+        private List<String> clientScopesGranted;
+        private List<String> additionalGrants;
+        private Map<String, List<ClientRoleRepresentation>> clientRoles;
+
+        public ApplicationRepresentation() {
+        }
+
+        public ApplicationRepresentation(ApplicationsBean.ApplicationEntry entry) {
```
Comment (abstractj) [inline services/src/main/java/org/keycloak/services/resources/account/ApplicationResource.java]: @ssilvert I just wonder if these representations would make more sense inside `org.keycloak.representations.account`. What do you think?

**Comment by abstractj** [comment]
@douglaspalmer that's not a blocker, but I added this to the Jira description https://issues.jboss.org/browse/KEYCLOAK-5628. We might need `inUse` and maybe a field with `grantedOn`. If we look at the wireframes, that's part of the screen.

Please, let me know what do you think.

File: services/src/main/java/org/keycloak/services/resources/account/ApplicationResource.java (hunk 3 above)
Comment (ssilvert) [reply services/src/main/java/org/keycloak/services/resources/account/ApplicationResource.java]: @abstractj I agree.  We need to be consistent.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/account/ApplicationRestServiceTest.java [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    public void testGetApplications() throws IOException {
+        TokenUtil viewToken = new TokenUtil("alice", "password");
+        oauth.doLogin("alice", "password");
+
+        List<ApplicationResource.ApplicationRepresentation> sessions = SimpleHttp
+                .doGet(getAccountUrl("applications"), httpClient).auth(viewToken.getToken())
+                .asJson(new TypeReference<List<ApplicationResource.ApplicationRepresentation>>() {
+                });
+
+        //TODO: assert response
```
Comment (ssilvert) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/account/ApplicationRestServiceTest.java]: Looks like this test is incomplete?

**Comment by ssilvert** [comment]
I don't see where the "Remove Access" endpoint is implemented.

**Review by ssilvert — CHANGES_REQUESTED** [review]

File: services/src/main/java/org/keycloak/services/resources/account/AccountRestService.java [hunk 5]
```diff
@@ -257,7 +258,7 @@ public Response getApplications() {
         List<ClientModel> clients = realm.getClients();
```
Comment (abstractj) [inline services/src/main/java/org/keycloak/services/resources/account/AccountRestService.java]: @douglaspalmer I'm really confused about this part. Are we getting all the clients from the `realm` or the clients which user authorized to act on their behalf? From what I tested here, seems to be the first case.

File: services/src/main/java/org/keycloak/services/resources/account/ApplicationResource.java [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+import org.jboss.resteasy.annotations.cache.NoCache;
+import org.jboss.resteasy.spi.HttpRequest;
+import org.keycloak.forms.account.freemarker.model.ApplicationsBean;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.RealmModel;
+import org.keycloak.models.UserModel;
+import org.keycloak.representations.idm.ClientRepresentation;
+import org.keycloak.services.managers.Auth;
+
+public class ApplicationResource {
```
Comment (abstractj) [inline services/src/main/java/org/keycloak/services/resources/account/ApplicationResource.java]: @douglaspalmer this part of the code seems to retrieve applications based on user's session (which I believe is our requirement), but digging into sources only integration tests are using it. Wouldn't be the case to use it also inside the applications endpoint?

**Comment by abstractj** [comment]
@douglaspalmer could you please follow up on the comments above?

**Review by abstractj — COMMENTED** [review]
Ignore the comments here, was added by accident and some how I cannot delete my own comments from GH.

**Comment by abstractj** [comment]
Superseded by https://github.com/keycloak/keycloak/pull/6382

**Pull request closed without merging**